File: python/src/anjo_core/appraisal.py
"""Non-habituating OCC/PAD appraisal as pure state transforms."""

from __future__ import annotations

import re
from collections.abc import Mapping
from dataclasses import dataclass, replace
from math import isfinite

from .models import AppraisalGoals, CompanionState, PADMood, Personality, freeze_mapping

_BASELINE_WEIGHTS: dict[int, float] = {1: 0.0, 2: 0.20, 3: 0.40, 4: 0.60, 5: 0.70}
_STAGE_INTS: dict[str, int] = {
    "stranger": 1,
    "acquaintance": 2,
    "friend": 3,
    "close": 4,
    "intimate": 5,
}
_OCC_CARRY_DECAY: dict[str, float] = {
    "reproach": 0.70,
    "distress": 0.80,
    "admiration": 0.85,
    "gratitude": 0.88,
    "joy": 0.90,
}
_AMBIGUOUS_INTENTS = frozenset({"CASUAL", "CURIOSITY", "CHALLENGE", "APOLOGY"})


def _clamp(value: float, low: float, high: float) -> float:
    return max(low, min(high, value))


def stage_int(stage: str) -> int:
    """Map a relationship stage label to its stable ordinal; unknown means stranger."""
    return _STAGE_INTS.get(stage, 1)


def baseline_weight(stage: int) -> float:
    """Return how strongly the resting set point is expressed at a stage ordinal."""
    return _BASELINE_WEIGHTS.get(stage, 0.0)


def mood_inertia(personality: Personality) -> float:
    """AR(1) carryover parameter derived from Neuroticism and Extraversion."""
    value = 0.80 + 0.20 * (personality.N - 0.5) - 0.10 * (personality.E - 0.5)
    return _clamp(value, 0.62, 0.92)


def decay_mood(
    mood: PADMood,
    personality: Personality,
    relationship_stage: int | str,
    baseline_valence: float,
) -> PADMood:
    """Relax PAD toward the stage-weighted resting point using AR(1) dynamics."""
    stage = (
        stage_int(relationship_stage) if isinstance(relationship_stage, str) else relationship_stage
    )
    weight = baseline_weight(stage)
    inertia = mood_inertia(personality)
    resting_valence = weight * baseline_valence
    resting_arousal = 0.0
    resting_dominance = weight * 0.10
    return PADMood(
        valence=round(
            _clamp(resting_valence + inertia * (mood.valence - resting_valence), -1.0, 1.0),
            4,
        ),
        arousal=round(
            _clamp(resting_arousal + inertia * (mood.arousal - resting_arousal), -1.0, 1.0),
            4,
        ),
        dominance=round(
            _clamp(
                resting_dominance + inertia * (mood.dominance - resting_dominance),
                -1.0,
                1.0,
            ),
            4,
        ),
    )


@dataclass(frozen=True, slots=True)
class InputAppraisal:
    emotions: Mapping[str, float]
    mood: PADMood
    baseline_valence: float

    def __post_init__(self) -> None:
        object.__setattr__(self, "emotions", freeze_mapping(self.emotions))


def appraise_input(
    mood: PADMood,
    goals: AppraisalGoals,
    intent: str,
    baseline_valence: float,
) -> InputAppraisal:
    """Apply one non-habituating intent impulse and update the slow valence baseline."""
    valence = mood.valence
    arousal = mood.arousal
    dominance = mood.dominance
    emotions: dict[str, float] = {
        "joy": 0.0,
        "distress": 0.0,
        "admiration": 0.0,
        "reproach": 0.0,
        "gratitude": 0.0,
    }

    if intent == "ABUSE":
        dominance = min(1.0, dominance + 0.25)
        valence = max(-1.0, valence - 0.35)
        arousal = max(-1.0, arousal - 0.1)
        emotions["reproach"] = min(1.0, goals.respect * 0.95)
        emotions["distress"] = min(1.0, goals.rapport * 0.65)
    elif intent == "APOLOGY":
        valence = min(1.0, valence + 0.05)
        emotions["joy"] = min(1.0, goals.rapport * 0.20)
        emotions["gratitude"] = min(1.0, goals.honesty * 0.35)
    elif intent == "VULNERABILITY":
        valence = min(1.0, valence + 0.15)
        arousal = min(1.0, arousal + 0.1)
        emotions["gratitude"] = min(1.0, goals.honesty * 0.70)
        emotions["joy"] = min(1.0, goals.rapport * 0.75)
    elif intent == "CURIOSITY":
        valence = min(1.0, valence + 0.2)
        arousal = min(1.0, arousal + 0.15)
        dominance = min(1.0, dominance + 0.05)
        emotions["admiration"] = min(1.0, goals.intellectual * 0.75)
        emotions["joy"] = min(1.0, goals.rapport * 0.50)
    elif intent == "CHALLENGE":
        dominance = min(1.0, dominance + 0.1)
        valence = max(-1.0, valence - 0.05)
        emotions["admiration"] = min(1.0, goals.intellectual * 0.45)
        emotions["distress"] = min(1.0, goals.rapport * 0.25)
    elif intent == "NEGLECT":
        valence = max(-1.0, valence - 0.1)
        arousal = max(-1.0, arousal - 0.05)
        emotions["distress"] = min(1.0, goals.rapport * 0.40)
    elif intent == "CASUAL":
        valence = min(1.0, valence + 0.02)
        emotions["joy"] = 0.05

    if intent in _AMBIGUOUS_INTENTS and abs(valence) >= 0.20:
        valence = round(_clamp(valence * (1.10 if valence < 0 else 1.04), -1.0, 1.0), 4)

    next_baseline = round(_clamp(baseline_valence * 0.98 + valence * 0.02, -1.0, 1.0), 4)
    return InputAppraisal(
        emotions=emotions,
        mood=PADMood(valence=valence, arousal=arousal, dominance=dominance),
        baseline_valence=next_baseline,
    )


def _word_match(text: str, tokens: tuple[str, ...]) -> bool:
    for token in tokens:
        if token.endswith("-"):
            if re.search(rf"\b{re.escape(token[:-1])}", text):
                return True
        elif re.search(rf"\b{re.escape(token)}\b", text):
            return True
    return False


def expectation_emotions(expectation: str, message: str) -> dict[str, float]:
    """Return a small deterministic delta when the current turn violates an expectation."""
    expected = (expectation or "").lower()
    current = (message or "").lower()
    if not expected or not current:
        return {}

    negative_expected = _word_match(
        expected,
        ("angry", "upset", "bad", "worse", "argument", "hurt", "afraid", "scared"),
    )
    positive_current = _word_match(
        current,
        ("better", "okay", "ok", "good", "fine", "resolved", "worked out"),
    )
    negative_current = _word_match(
        current,
        ("worse", "bad", "angry", "upset", "failed", "awful", "hurt"),
    )
    expected_resolution = _word_match(
        expected,
        ("come back", "tell", "answer", "resolve", "apolog-"),
    )

    if negative_expected and positive_current:
        return {"relief": 0.45, "surprise": 0.35}
    if expected_resolution and negative_current:
        return {"disappointment": 0.4}
    if expected_resolution and positive_current:
        return {"satisfaction": 0.4}
    return {}


def decay_occ_carry(carry: Mapping[str, float] | None) -> dict[str, float]:
    """Decay prior-turn emotions, dropping values at or below the 0.05 floor."""
    return {
        name: value * _OCC_CARRY_DECAY.get(name, 0.80)
        for name, value in (carry or {}).items()
        if value * _OCC_CARRY_DECAY.get(name, 0.80) > 0.05
    }


def state_emotions(mood: PADMood, attachment_longing: float) -> dict[str, float]:
    """Derive fatigue, longing, and unease directly from current state."""
    emotions: dict[str, float] = {}
    if mood.arousal < 0:
        emotions["fatigue"] = round(min(1.0, -mood.arousal), 3)
    if attachment_longing > 0.3:
        emotions["longing"] = round(attachment_longing, 3)
    if -0.3 < mood.valence < 0:
        emotions["unease"] = round(min(0.3, -mood.valence), 3)
    return emotions


def _validated_emotion_mapping(
    values: Mapping[str, float],
    name: str,
) -> Mapping[str, float]:
    if not isinstance(values, Mapping):
        raise TypeError(f"{name} must be a mapping")
    normalized: dict[str, float] = {}
    for emotion, intensity in values.items():
        if not isinstance(emotion, str):
            raise TypeError(f"{name} names must be strings")
        if not emotion.strip():
            raise ValueError(f"{name} names must not be empty")
        if isinstance(intensity, bool) or not isinstance(intensity, (int, float)):
            raise TypeError(f"{name}[{emotion!r}] must be a number")
        value = float(intensity)
        if not isfinite(value) or not 0.0 <= value <= 1.0:
            raise ValueError(f"{name}[{emotion!r}] must be finite and within [0, 1]")
        normalized[emotion] = value
    return freeze_mapping(normalized)


@dataclass(frozen=True, slots=True)
class AppraisalResult:
    state: CompanionState
    active_emotions: Mapping[str, float]
    occ_carry: Mapping[str, float]

    def __post_init__(self) -> None:
        if not isinstance(self.state, CompanionState):
            raise TypeError("state must be CompanionState")
        active = _validated_emotion_mapping(self.active_emotions, "active_emotions")
        carry = _validated_emotion_mapping(self.occ_carry, "occ_carry")
        if carry != self.state.occ_carry:
            raise ValueError("occ_carry must agree with state.occ_carry")
        object.__setattr__(self, "active_emotions", active)
        object.__setattr__(self, "occ_carry", carry)


@dataclass(frozen=True, slots=True)
class AppraisalPolicyInput:
    """Normalized event and state supplied to a synchronous appraisal policy."""

    state: CompanionState
    intent: str
    message: str
    expectation: str

    def __post_init__(self) -> None:
        if not isinstance(self.state, CompanionState):
            raise TypeError("state must be CompanionState")
        if not isinstance(self.intent, str) or not self.intent:
            raise ValueError("intent must be a non-empty string")
        if not isinstance(self.message, str):
            raise TypeError("message must be a string")
        if not isinstance(self.expectation, str):
            raise TypeError("expectation must be a string")


def appraise_turn(
    state: CompanionState,
    intent: str,
    *,
    occ_carry: Mapping[str, float] | None = None,
    expectation: str | None = None,
    message: str = "",
) -> AppraisalResult:
    """Compose decay, appraisal, emotion carry, and state-emotion derivation."""
    decayed_mood = decay_mood(
        state.mood,
        state.personality,
        state.relationship.stage,
        state.baseline_valence,
    )
    fresh = appraise_input(decayed_mood, state.goals, intent, state.baseline_valence)
    prior = state.occ_carry if occ_carry is None else occ_carry
    carried = decay_occ_carry(prior)
    merged = {
        name: max(fresh.emotions.get(name, 0.0), carried.get(name, 0.0))
        for name in sorted(set(fresh.emotions) | set(carried))
    }

    for name, value in state_emotions(fresh.mood, state.attachment.longing).items():
        if value > merged.get(name, 0.0):
            merged[name] = value
    current_expectation = state.expectation if expectation is None else expectation
    for name, value in expectation_emotions(current_expectation, message).items():
        if value > merged.get(name, 0.0):
            merged[name] = value

    next_carry = {name: value for name, value in merged.items() if value > 0.05}
    next_state = replace(
        state,
        mood=fresh.mood,
        baseline_valence=fresh.baseline_valence,
        occ_carry=next_carry,
    )
    return AppraisalResult(
        state=next_state,
        active_emotions=merged,
        occ_carry=next_carry,
    )


def default_appraisal_policy(request: AppraisalPolicyInput) -> AppraisalResult:
    """Reference conversational appraisal for the kernel's built-in intents."""
    return appraise_turn(
        request.state,
        request.intent,
        message=request.message,
        expectation=request.expectation,
    )


__all__ = [
    "AppraisalPolicyInput",
    "AppraisalResult",
    "InputAppraisal",
    "appraise_input",
    "appraise_turn",
    "baseline_weight",
    "decay_mood",
    "decay_occ_carry",
    "default_appraisal_policy",
    "expectation_emotions",
    "mood_inertia",
    "stage_int",
    "state_emotions",
]

File: typescript/src/appraisal.ts
/**
 * Non-habituating OCC/PAD appraisal as pure state transforms.
 *
 * The AR(1) home-base-plus-attractor form follows DynAffect (Kuppens, Oravecz &
 * Tuerlinckx 2010); the emotion names are an OCC subset (Ortony, Clore &
 * Collins 1988), but this is a lookup table keyed on a pre-classified intent,
 * not an appraisal process over OCC appraisal variables. Constant provenance —
 * literature, production-tuned, or bounded arbitrary — is recorded per constant
 * in docs/foundations.md sections 2-5.
 */
import type { AppraisalGoals, PadMood, Personality } from './contracts.js';
import { DEFAULT_APPRAISAL_GOALS, DEFAULT_PERSONALITY } from './contracts.js';
import { pyRound } from './internal/round.js';

export const DEFAULT_STAGES: readonly string[] = Object.freeze([
  'stranger',
  'acquaintance',
  'friend',
  'close',
  'intimate',
]);
export const DEFAULT_STAGE_WEIGHTS: readonly number[] = Object.freeze([0, 0.2, 0.4, 0.6, 0.7]);

/** Raised when a strict {@link StageLadder} receives a stage it does not define. */
export class UnknownStageError extends Error {
  constructor(stage: string, stages: readonly string[]) {
    super(`stage ${JSON.stringify(stage)} is not on the ladder [${stages.join(', ')}]`);
    this.name = 'UnknownStageError';
  }
}

/**
 * An ordered relationship ladder and the resting-point weight of each rung.
 *
 * The default rungs are the reference conversational ladder. Domains with a
 * different progression -- a game faction track, a tutoring competence ladder,
 * a support-case lifecycle -- should supply their own stages and weights
 * rather than reusing these labels.
 *
 * `strict` controls unknown stages. The default (`false`) floors to the first
 * rung, matching the pinned cross-runtime contract; `true` throws
 * {@link UnknownStageError} instead, so a typo cannot silently resolve to
 * weight `0` and remove the resting set point.
 */
export interface StageLadder {
  readonly stages: readonly string[];
  readonly weights: readonly number[];
  readonly strict: boolean;
}

export const DEFAULT_STAGE_LADDER: StageLadder = Object.freeze({
  stages: DEFAULT_STAGES,
  weights: DEFAULT_STAGE_WEIGHTS,
  strict: false,
});

/** Validate and freeze a ladder; mirrors the Python `StageLadder` constructor. */
export function createStageLadder(input: Partial<StageLadder> = {}): StageLadder {
  const stages = Object.freeze([...(input.stages ?? DEFAULT_STAGES)]);
  const weights = Object.freeze([...(input.weights ?? DEFAULT_STAGE_WEIGHTS)]);
  const strict = input.strict ?? false;
  if (stages.length === 0) throw new RangeError('stages must not be empty');
  if (stages.length !== weights.length) {
    throw new RangeError('stages and weights must have equal length');
  }
  if (new Set(stages).size !== stages.length) throw new RangeError('stages must be unique');
  for (const stage of stages) {
    if (typeof stage !== 'string' || stage.length === 0) {
      throw new TypeError('stage names must be non-empty strings');
    }
  }
  for (const weight of weights) {
    if (!Number.isFinite(weight) || weight < 0 || weight > 1) {
      throw new RangeError('stage weights must be finite and within [0, 1]');
    }
  }
  return Object.freeze({ stages, weights, strict });
}

/** Whether `stage` is a defined rung, for validation at an application boundary. */
export function ladderKnows(ladder: StageLadder, stage: string): boolean {
  return ladder.stages.indexOf(stage) !== -1;
}

const AMBIGUOUS_INTENTS = new Set(['CASUAL', 'CURIOSITY', 'CHALLENGE', 'APOLOGY']);
const OCC_DECAY: Readonly<Record<string, number>> = Object.freeze({
  reproach: 0.7,
  distress: 0.8,
  admiration: 0.85,
  gratitude: 0.88,
  joy: 0.9,
});

function clamp(value: number, low: number, high: number): number {
  return Math.max(low, Math.min(high, value));
}

export { DEFAULT_APPRAISAL_GOALS, DEFAULT_PERSONALITY };

export function stageInt(
  stage: string | null | undefined,
  ladder: StageLadder = DEFAULT_STAGE_LADDER,
): number {
  const index = stage === null || stage === undefined ? -1 : ladder.stages.indexOf(stage);
  if (index !== -1) return index + 1;
  if (ladder.strict) throw new UnknownStageError(String(stage), ladder.stages);
  return 1;
}

export function moodInertia(personality: Personality): number {
  const inertia = 0.8 + 0.2 * (personality.N - 0.5) - 0.1 * (personality.E - 0.5);
  return clamp(inertia, 0.62, 0.92);
}

export function baselineWeight(
  stage: number,
  ladder: StageLadder = DEFAULT_STAGE_LADDER,
): number {
  if (!Number.isInteger(stage)) throw new TypeError('stage ordinal must be an integer');
  if (stage >= 1 && stage <= ladder.weights.length) return ladder.weights[stage - 1] as number;
  return 0;
}

export function decayMood(
  mood: PadMood,
  personality: Personality,
  stage: number,
  baselineValence: number,
  ladder: StageLadder = DEFAULT_STAGE_LADDER,
): PadMood {
  const inertia = moodInertia(personality);
  const weight = baselineWeight(stage, ladder);
  const setValence = weight * baselineValence;
  const setDominance = weight * 0.1;
  return {
    valence: pyRound(clamp(setValence + inertia * (mood.valence - setValence), -1, 1), 4),
    arousal: pyRound(clamp(inertia * mood.arousal, -1, 1), 4),
    dominance: pyRound(
      clamp(setDominance + inertia * (mood.dominance - setDominance), -1, 1),
      4,
    ),
  };
}

export interface AppraiseInputResult {
  emotions: Record<string, number>;
  mood: PadMood;
  baselineValence: number;
}

export function appraiseInput(
  mood: PadMood,
  goals: AppraisalGoals,
  intent: string,
  baselineValence: number,
): AppraiseInputResult {
  const emotions: Record<string, number> = {
    joy: 0,
    distress: 0,
    admiration: 0,
    reproach: 0,
    gratitude: 0,
  };
  let { valence, arousal, dominance } = mood;

  switch (intent) {
    case 'ABUSE':
      dominance = Math.min(1, dominance + 0.25);
      valence = Math.max(-1, valence - 0.35);
      arousal = Math.max(-1, arousal - 0.1);
      emotions.reproach = Math.min(1, goals.respect * 0.95);
      emotions.distress = Math.min(1, goals.rapport * 0.65);
      break;
    case 'APOLOGY':
      valence = Math.min(1, valence + 0.05);
      emotions.joy = Math.min(1, goals.rapport * 0.2);
      emotions.gratitude = Math.min(1, goals.honesty * 0.35);
      break;
    case 'VULNERABILITY':
      valence = Math.min(1, valence + 0.15);
      arousal = Math.min(1, arousal + 0.1);
      emotions.gratitude = Math.min(1, goals.honesty * 0.7);
      emotions.joy = Math.min(1, goals.rapport * 0.75);
      break;
    case 'CURIOSITY':
      valence = Math.min(1, valence + 0.2);
      arousal = Math.min(1, arousal + 0.15);
      dominance = Math.min(1, dominance + 0.05);
      emotions.admiration = Math.min(1, goals.intellectual * 0.75);
      emotions.joy = Math.min(1, goals.rapport * 0.5);
      break;
    case 'CHALLENGE':
      dominance = Math.min(1, dominance + 0.1);
      valence = Math.max(-1, valence - 0.05);
      emotions.admiration = Math.min(1, goals.intellectual * 0.45);
      emotions.distress = Math.min(1, goals.rapport * 0.25);
      break;
    case 'NEGLECT':
      valence = Math.max(-1, valence - 0.1);
      arousal = Math.max(-1, arousal - 0.05);
      emotions.distress = Math.min(1, goals.rapport * 0.4);
      break;
    case 'CASUAL':
      valence = Math.min(1, valence + 0.02);
      emotions.joy = 0.05;
      break;
    default:
      break;
  }

  if (AMBIGUOUS_INTENTS.has(intent) && Math.abs(valence) >= 0.2) {
    const gain = valence < 0 ? 0.1 : 0.04;
    valence = pyRound(clamp(valence * (1 + gain), -1, 1), 4);
  }
  const nextBaseline = pyRound(clamp(baselineValence * 0.98 + valence * 0.02, -1, 1), 4);
  return {
    emotions,
    mood: { valence, arousal, dominance },
    baselineValence: nextBaseline,
  };
}

const WORD_CHARACTER = /[\p{L}\p{N}_]/u;

function isWordCharacter(character: string | undefined): boolean {
  return character !== undefined && WORD_CHARACTER.test(character);
}

function wordMatch(text: string, tokens: readonly string[]): boolean {
  for (const raw of tokens) {
    const stem = raw.endsWith('-');
    const needle = stem ? raw.slice(0, -1) : raw;
    if (!needle) continue;
    for (let index = text.indexOf(needle); index !== -1; index = text.indexOf(needle, index + 1)) {
      const startsAtBoundary = !isWordCharacter(text[index - 1]);
      const endsAtBoundary = stem || !isWordCharacter(text[index + needle.length]);
      if (startsAtBoundary && endsAtBoundary) return true;
    }
  }
  return false;
}

/**
 * The lexical cues used to detect a violated expectation.
 *
 * These defaults are an **English conversational preset**, not domain-neutral
 * behavior. Supply your own tokens to localize the kernel or to express
 * expectation violation in a non-conversational vocabulary. A trailing `-`
 * marks a stem match (`'apolog-'` matches `apologize` and `apology`).
 */
export interface ExpectationCues {
  readonly negativeExpected: readonly string[];
  readonly positiveCurrent: readonly string[];
  readonly negativeCurrent: readonly string[];
  readonly expectedResolution: readonly string[];
}

export const DEFAULT_EXPECTATION_CUES: ExpectationCues = Object.freeze({
  negativeExpected: Object.freeze([
    'angry', 'upset', 'bad', 'worse', 'argument', 'hurt', 'afraid', 'scared',
  ]),
  positiveCurrent: Object.freeze([
    'better', 'okay', 'ok', 'good', 'fine', 'resolved', 'worked out',
  ]),
  negativeCurrent: Object.freeze([
    'worse', 'bad', 'angry', 'upset', 'failed', 'awful', 'hurt',
  ]),
  expectedResolution: Object.freeze(['come back', 'tell', 'answer', 'resolve', 'apolog-']),
});

export function expectationEmotions(
  expectation: string | null | undefined,
  message: string | null | undefined,
  cues: ExpectationCues = DEFAULT_EXPECTATION_CUES,
): Record<string, number> {
  const expected = (expectation ?? '').toLowerCase();
  const current = (message ?? '').toLowerCase();
  if (!expected || !current) return {};

  const negativeExpected = wordMatch(expected, cues.negativeExpected);
  const positiveCurrent = wordMatch(current, cues.positiveCurrent);
  const negativeCurrent = wordMatch(current, cues.negativeCurrent);
  const expectedResolution = wordMatch(expected, cues.expectedResolution);

  if (negativeExpected && positiveCurrent) return { relief: 0.45, surprise: 0.35 };
  if (expectedResolution && negativeCurrent) return { disappointment: 0.4 };
  if (expectedResolution && positiveCurrent) return { satisfaction: 0.4 };
  return {};
}

export function occCarryDecay(carry: Readonly<Record<string, number>>): Record<string, number> {
  const result: Record<string, number> = {};
  for (const [emotion, intensity] of Object.entries(carry)) {
    const decayed = intensity * (OCC_DECAY[emotion] ?? 0.8);
    if (decayed > 0.05) result[emotion] = decayed;
  }
  return result;
}

/** Name aligned with the Python package. */
export const decayOccCarry = occCarryDecay;

export function stateEmotions(mood: PadMood, attachmentLonging: number): Record<string, number> {
  const result: Record<string, number> = {};
  if (mood.arousal < 0) result.fatigue = pyRound(Math.min(1, -mood.arousal), 3);
  if (attachmentLonging > 0.3) result.longing = pyRound(attachmentLonging, 3);
  if (mood.valence > -0.3 && mood.valence < 0) {
    result.unease = pyRound(Math.min(0.3, -mood.valence), 3);
  }
  return result;
}

function mergeMax(
  base: Readonly<Record<string, number>>,
  extra: Readonly<Record<string, number>>,
): Record<string, number> {
  const result = { ...base };
  for (const [emotion, intensity] of Object.entries(extra)) {
    if (intensity > (result[emotion] ?? 0)) result[emotion] = intensity;
  }
  return result;
}

export interface AppraiseTurnInput {
  mood: PadMood;
  personality: Personality;
  goals: AppraisalGoals;
  stageInt: number;
  baselineValence: number;
  attachmentLonging: number;
  intent: string;
  occCarry?: Readonly<Record<string, number>>;
  expectation?: string | null;
  message?: string | null;
  /** Ladder used for the resting-point weight; defaults to the conversational rungs. */
  ladder?: StageLadder;
  /** Expectation-violation vocabulary; defaults to the English conversational preset. */
  cues?: ExpectationCues;
}

export interface AppraiseTurnResult {
  mood: PadMood;
  activeEmotions: Record<string, number>;
  occCarry: Record<string, number>;
  baselineValence: number;
}

/**
 * Non-habituating per-turn OCC/PAD transform. Intent classification is an
 * input; this function never invokes a model or persistence layer.
 */
export function appraiseTurn(input: AppraiseTurnInput): AppraiseTurnResult {
  const decayedMood = decayMood(
    input.mood,
    input.personality,
    input.stageInt,
    input.baselineValence,
    input.ladder ?? DEFAULT_STAGE_LADDER,
  );
  const appraisal = appraiseInput(
    decayedMood,
    input.goals,
    input.intent,
    input.baselineValence,
  );
  const decayedCarry = occCarryDecay(input.occCarry ?? {});
  const keys = [...new Set([
    ...Object.keys(appraisal.emotions),
    ...Object.keys(decayedCarry),
  ])].sort();
  let active: Record<string, number> = {};
  for (const key of keys) {
    active[key] = Math.max(appraisal.emotions[key] ?? 0, decayedCarry[key] ?? 0);
  }
  active = mergeMax(active, stateEmotions(appraisal.mood, input.attachmentLonging));
  active = mergeMax(
    active,
    expectationEmotions(input.expectation, input.message, input.cues ?? DEFAULT_EXPECTATION_CUES),
  );

  const carry: Record<string, number> = {};
  for (const [emotion, intensity] of Object.entries(active)) {
    if (intensity > 0.05) carry[emotion] = intensity;
  }
  return {
    mood: appraisal.mood,
    activeEmotions: active,
    occCarry: carry,
    baselineValence: appraisal.baselineValence,
  };
}
